Libraries/LibGC/Heap.h

```diff
@@ -73,13 +73,13 @@ class Heap : public HeapBase {
 
     void uproot_cell(Cell* cell);
 
+    void defer_gc();
+    void undefer_gc();
+    bool is_gc_deferred() const { return m_gc_deferrals > 0; }
+
 private:
     friend class MarkingVisitor;
     friend class GraphConstructorVisitor;
-    friend class DeferGC;
```

Comment:
I'm not sure if anything is, but it seems a lot more natural to call the APIs directly from swift than to use the DeferGC object. I'm worried that Swift might see that the value is unused after the declaration and call the dtor right then and there :shrug: . I believe Gabor mentioned this type of issue in a recent Swift forums post.

Reply:
With copyable types, Swift doesn't follow strict scoping for destruction of values. Rather, Swift's optimizer can eagerly destroy values that are no longer used, including (for example) doing a destructive move rather than a copy when the source isn't used any more. This is an important optimization for reference counting, because it lets us avoid extraneous ref-count traffic, but means that you don't get the strict scoping that RAII depends on.

Swift's noncopyable types are more amenable to RAII. A local variable of noncopyable type will be destroyed either when it is consumed (i.e., passed to a `consuming` parameter or function, after which it can't be touched) or at the end of its scope. It's possible to force a given C++ type to be imported as noncopyable using the `SWIFT_NONCOPYABLE` macro from `<swift/ bridging>`, making it same to use the RAII idiom in Swift.

Reply:
That works, thanks for the explaination! It was a bit surprising that the compiler treats 

```
let _ = GC.DeferGC(self)
```

differently from

```
let deferRAII = GC.DeferGC(self)
_ = deferRAII
```

The second form does what I'd expect, while the first is eliminated as completely unused.